checker/src/org/checkerframework/checker/lock/LockAnnotatedTypeFactory.java

```diff
@@ -96,28 +107,310 @@ public LockTransfer createFlowTransferFunction(CFAbstractAnalysis<CFValue, LockS
+            this.annotationClass=annotationClass;
+        }
+
+        public String getNameOfSideEffectAnnotation() {
+            return annotation;
+        }
+
+        public Class<? extends Annotation> getAnnotationClass() {
+            return annotationClass;
+        }
+
+        /**
+         * Returns true if the receiver side effect annotation is weaker
+         * than side effect annotation 'other'.
+         */
+        boolean isWeakerThan(SideEffectAnnotation other) {
+            boolean weaker = false;
+
+            switch (other) {
+                case MAYRELEASELOCKS:
+                    break;
+                case RELEASESNOLOCKS:
+                    if (this == SideEffectAnnotation.MAYRELEASELOCKS) {
+                        weaker = true;
+                    }
+                    break;
+                case LOCKINGFREE:
+                    switch (this) {
+                        case MAYRELEASELOCKS:
+                        case RELEASESNOLOCKS:
+                            weaker = true;
+                            break;
+                        default:
+                    }
+                    break;
+                case SIDEEFFECTFREE:
+                    switch (this) {
+                        case MAYRELEASELOCKS:
+                        case RELEASESNOLOCKS:
+                        case LOCKINGFREE:
+                            weaker = true;
+                            break;
+                        default:
+                    }
+                    break;
+                case PURE:
+                    switch (this) {
+                        case MAYRELEASELOCKS:
+                        case RELEASESNOLOCKS:
+                        case LOCKINGFREE:
+                        case SIDEEFFECTFREE:
+                            weaker = true;
+                            break;
+                        default:
+                    }
+                    break;
+            }
+
+            return weaker;
+        }
+
+        static SideEffectAnnotation weakest = null;
+        public static SideEffectAnnotation weakest() {
+            if(weakest == null) {
+                for (SideEffectAnnotation sea : SideEffectAnnotation.values()) {
+                    if (weakest == null) weakest = sea;
+                    if (sea.isWeakerThan(weakest)) {
+                        weakest = sea;
+                    }
+                }
+            }
+            return weakest;
+        }
     }
 
-    class LockQualifierHierarchy extends GraphQualifierHierarchy {
+    // Indicates which side effect annotation is present on the given method.
+    // If more than one annotation is present, this method issues an error (if issueErrorIfMoreThanOnePresent is true)
+    // and returns the annotation providing the weakest guarantee.
+    // If no annotation is present, return RELEASESNOLOCKS as the default, and MAYRELEASELOCKS
+    // as the default for unannotated code.
```

Comment:
Only when visiting the method definition, not a method invocation. This prevents multiple error messages for the same method. Adding a comment.